docs/rest/keysend.md

```diff
@@ -0,0 +1,122 @@
+from secrets import token_hex
+from typing import Tuple
+
+import httpx
+
+
+
+def get_lnd_headers_cert(
+    admin: bool = False, local: bool = False, node: str = None
+) -> Tuple[dict, str]:
+    """Return the headers and certificate for connecting, if macaroon passed as string
+    does not return a certificate (Voltage)"""
+    if not node:
+        node = Config.LOCAL_LND_NODE_ADDRESS
+
+    # maintain option to work with local macaroon and umbrel
+    macaroon_folder = ".macaroon"
+    if not admin:
+        macaroon_file = "invoices.macaroon"
+    else:
+        macaroon_file = "admin.macaroon"
+    macaroon_path = os.path.join(macaroon_folder, macaroon_file)
+    cert = os.path.join(macaroon_folder, "tls.cert")
+    macaroon = codecs.encode(open(macaroon_path, "rb").read(), "hex")
+    headers = {"Grpc-Metadata-macaroon": macaroon}
+    return headers, cert
+
+
+def b64_hex_transform(plain_str: str) -> str:
+    """Returns the b64 transformed version of a hex string"""
+    a_string = bytes.fromhex(plain_str)
+    return base64.b64encode(a_string).decode()
+
+
+def b64_transform(plain_str: str) -> str:
+    """Returns the b64 transformed version of a string"""
+    return base64.b64encode(plain_str.encode()).decode()
+
+
+def send_keysend(
+    amt: int,
+    dest_pubkey: str = "",
+    hive_accname: str = "brianoflondon",
+) -> dict:
+    """Pay a keysend invoice using the chosen node"""
+    node = "https://umbrel.local:8080/"
+    headers, cert = get_lnd_headers_cert(admin=True, node=node)
+    if not dest_pubkey:
+        dest_pubkey = my_voltage_public_key
+
+    # Base 64 encoded destination bytes
+    dest = b64_hex_transform(dest_pubkey)
+
+    # We generate a random 32 byte Hex pre_image here.
+    pre_image = token_hex(32)
+    # This is the hash of the pre-image
+    payment_hash = sha256(bytes.fromhex(pre_image))
+
+    # The record 5482373484 is special: it carries the pre_image
+    # to the destination so it can be compared with the hash we
+    # pass via the payment_hash
+    dest_custom_records = {
+        5482373484: b64_hex_transform(pre_image),
+        818818: b64_transform(hive_accname),
+    }
+
+    url = f"{node}v1/channels/transactions"
+    data = {
+        "dest": dest,
+        "amt": amt,
+        "payment_hash": b64_hex_transform(payment_hash.hexdigest()),
+        "dest_custom_records": dest_custom_records,
+    }
+
+    response = httpx.post(
+        url=url, headers=headers, data=json.dumps(data), verify=cert
+        )
+
+    print(json.dumps(response.json(), indent=2))
+    return json.dumps(response.json(), indent=2)
+
+```
+
+
+This explanation of what is going on here is pretty useful too:
+
+>Since you're not paying an invoice, the receiver doesn't know the preimage for the given payment's hash. You need to add the preimage to your custom records. The number is 5482373484.
+>
+>It also looks like you need to actually set the payment's hash (payment_hash=sha256(preimage)) of the payment.
```

Comment:
If you didn't set the hash to the preimage hash then the preimage wouldn't work to unlock the incoming HTLC on-chain and so you wouldn't receive funds, this applies across the route since all the HTLCs get locked to the hash and the outgoing HTLC is paid for by the incoming HTLC to the same preimage